### PR TITLE
Add support for non aggregated columns in `OrderedAggregate`

### PR DESCRIPTION
**Pull request by frouioui**
## Description

We do not keep track of the [ONLY_FULL_GROUP_BY](https://dev.mysql.com/doc/refman/5.7/en/sql-mode.html#sqlmode_only_full_group_by) setting in VTGate, thus we were failing certain query like the one below even though they should be executed if the setting is set to false.

```
select city, age from user group by city
# age is a non aggregated column here, the query should fail if sql_mode=only_full_group_by
```

This PR adds support for non aggregated columns in the `OrderedAggregate`. We should be able to plan a query that has non aggregated columns and send it to MySQL. Bases on the MySQL mode, MySQL can decide to fail this query or not.

## Related Issue(s)

- fixes https://github.com/vitessio/vitess/issues/9889

## Checklist

-   [ ] "Backport me!" label has been added if this change should be backported
-   [x] Tests were added or are not required
-   [x] Documentation was added or is not required

### CODE REVIEW

**Comment by frouioui** [comment]
This Pull Request depends on https://github.com/vitessio/vitess/pull/10163 for a fix in how we handle reserved connections for field queries. Do not merge until https://github.com/vitessio/vitess/pull/10163 is merged.

**Comment by derekperkins** [comment]
To clarify on the description, this will also allow for similar queries to succeed even if [ONLY_FULL_GROUP_BY](https://dev.mysql.com/doc/refman/5.7/en/sql-mode.html#sqlmode_only_full_group_by) is enabled, when MySQL can determine that those columns are functionally dependent on a grouping column

File: go/vt/vtgate/engine/ordered_aggregate.go [hunk 1]
```diff
@@ -102,8 +102,9 @@ type AggregateParams struct {
 	WAssigned   bool
 	CollationID collations.ID
 
-	Alias string `json:",omitempty"`
-	Expr  sqlparser.Expr
+	Alias    string `json:",omitempty"`
+	Expr     sqlparser.Expr
+	Original *sqlparser.AliasedExpr
```
Comment (systay) [inline go/vt/vtgate/engine/ordered_aggregate.go:107]: Why do we need `original`? The only uses I see should be possible to handle using the Expr field, right?

Comment (frouioui) [reply go/vt/vtgate/engine/ordered_aggregate.go:107]: `Expr` field is the AST node that contains the aggregate function (i.e. `count(id)`), whereas `Original` is simply `id`.

Comment (frouioui) [reply go/vt/vtgate/engine/ordered_aggregate.go:107]: `Expr` field is the AST node that contains the aggregate function (i.e. `count(id)`), whereas `Original` is simply `id`.

Comment (systay) [reply go/vt/vtgate/engine/ordered_aggregate.go:107]: Hmm... Now am I'm even more confused. Why is the expression wrapped in AliasedExpr?

File: go/vt/vtgate/planbuilder/aggregation_pushing.go [hunk 2]
```diff
@@ -28,17 +28,11 @@ import (
...
 
 // pushAggregation pushes grouping and aggregation as far down in the tree as possible
-func (hp *horizonPlanning) pushAggregation(
-	ctx *plancontext.PlanningContext,
-	plan logicalPlan,
-	grouping []abstract.GroupBy,
-	aggregations []abstract.Aggr,
-	ignoreOutputOrder bool,
-) (output logicalPlan, groupingOffsets []offsets, outputAggrsOffset [][]offsets, err error) {
+func (hp *horizonPlanning) pushAggregation(ctx *plancontext.PlanningContext, plan logicalPlan, grouping []abstract.GroupBy, aggregations []abstract.Aggr, ignoreOutputOrder bool) (output logicalPlan, groupingOffsets []offsets, outputAggrsOffset [][]offsets, err error) {
```
Comment (systay) [inline go/vt/vtgate/planbuilder/aggregation_pushing.go:31]: nit: these lines are difficult to read when in a single line. mind splitting it again?

File: go/vt/vtgate/planbuilder/aggregation_pushing.go [hunk 3]
```diff
@@ -373,7 +371,7 @@ func splitAggregationsToLeftAndRight(
 	var lhsAggrs, rhsAggrs []*abstract.Aggr
 	for _, aggr := range aggregations {
 		newAggr := aggr
-		if isCountStar(aggr.Func) {
+		if aggr.Func != nil && isCountStar(aggr.Func) {
```
Comment (systay) [inline go/vt/vtgate/planbuilder/aggregation_pushing.go:374]: nit: you could move in the nil checking into `isCountStar`

File: go/vt/vtgate/engine/ordered_aggregate.go [hunk 4]
```diff
@@ -476,6 +478,7 @@ func merge(
 			data, _ := proto.Marshal(vgtid)
 			val, _ := sqltypes.NewValue(sqltypes.VarBinary, data)
 			result[aggr.Col] = val
+		case AggregateRandom:
```
Comment (systay) [inline go/vt/vtgate/engine/ordered_aggregate.go:481]: this is correct, but also a little confusing. maybe add a comment explaining what's going on here?

File: go/vt/vtgate/planbuilder/abstract/queryprojection.go [hunk 5]
```diff
@@ -46,7 +46,6 @@ type (
 		OrderExprs         []OrderBy
 		CanPushDownSorting bool
 		HasStar            bool
-		ProjectionError    error
```
Comment (systay) [inline go/vt/vtgate/planbuilder/abstract/queryprojection.go:49]: 🎉

**Review by systay — APPROVED** [review]
LGTM

You probably need to merge main into this - the plantest outputs have changed.

**Comment by systay** [comment]
we should probably write something about this in our docs, so that people know that Vitess allows it and that the functionality depends on mysql to also allow it

File: go/vt/vtgate/engine/ordered_aggregate.go (hunk 1 above)
Comment (systay) [reply go/vt/vtgate/engine/ordered_aggregate.go:107]: Never mind. I get it now

**Comment by frouioui** [comment]
> we should probably write something about this in our docs, so that people know that Vitess allows it and that the functionality depends on mysql to also allow it

I feel like this PR is enhancing the compatibility with MySQL by letting MySQL handle more work. We are just mimicking the behavior that users expect when using MySQL.

**Pull request merged**